modyn/storage/src/internal/grpc/storage_service_impl.cpp

```diff
@@ -0,0 +1,591 @@
+  std::tie(start_index, limit) = get_partition_for_worker(request->worker_id(), request->total_workers(), total_keys);
+
+  std::vector<int64_t> keys;
+  soci::statement stmt = (session.prepare << "SELECT sample_id FROM Sample WHERE dataset_id = :dataset_id ORDER BY "
+                                             "sample_id OFFSET :start_index LIMIT :limit",
+                          soci::use(dataset_id), soci::use(start_index), soci::use(limit));
+  stmt.execute();
+
+  int64_t key_value;
+  stmt.exchange(soci::into(key_value));
+  while (stmt.fetch()) {
+    keys.push_back(key_value);
+  }
+
+  modyn::storage::GetDataPerWorkerResponse response;
+  for (auto key : keys) {
+    response.add_keys(key);
+    if (response.keys_size() % sample_batch_size_ == 0) {
+      writer->Write(response);
+      response.Clear();
+    }
+  }
+
+  if (response.keys_size() > 0) {
+    writer->Write(response);
+  }
+
+  return grpc::Status::OK;
+}
+
+std::tuple<int64_t, int64_t> StorageServiceImpl::get_partition_for_worker(int64_t worker_id, int64_t total_workers,
+                                                                          int64_t total_num_elements) {
+  if (worker_id < 0 || worker_id >= total_workers) {
+    throw std::invalid_argument("Asked for invalid worker id!");
+  }
+
+  int64_t subset_size = total_num_elements / total_workers;
+  int64_t worker_subset_size = subset_size;
+
+  int64_t threshold = total_num_elements % total_workers;
+  if (threshold > 0) {
+    if (worker_id < threshold) {
+      worker_subset_size += 1;
+      int64_t start_index = worker_id * (subset_size + 1);
+      return {start_index, worker_subset_size};
+    } else {
+      int64_t start_index = threshold * (subset_size + 1) + (worker_id - threshold) * subset_size;
+      return {start_index, worker_subset_size};
+    }
+  } else {
+    int64_t start_index = worker_id * subset_size;
+    return {start_index, worker_subset_size};
+  }
+}
+
+grpc::Status StorageServiceImpl::GetDatasetSize(grpc::ServerContext* context,
+                                                const modyn::storage::GetDatasetSizeRequest* request,
+                                                modyn::storage::GetDatasetSizeResponse* response) {
+  const StorageDatabaseConnection storage_database_connection = StorageDatabaseConnection(config_);
+  soci::session session = storage_database_connection.get_session();
+
+  // Check if the dataset exists
+  int64_t dataset_id = get_dataset_id(request->dataset_id(), session);
+
+  if (dataset_id == 0) {
+    SPDLOG_ERROR("Dataset {} does not exist.", request->dataset_id());
+    return {grpc::StatusCode::NOT_FOUND, "Dataset does not exist."};
+  }
+
+  int64_t total_keys = 0;
+  soci::statement count_stmt = (session.prepare << "SELECT COUNT(*) FROM Sample WHERE dataset_id = :dataset_id",
+                                soci::into(total_keys), soci::use(dataset_id));
+
+  count_stmt.execute();
+  response->set_num_keys(total_keys);
```

Comment:
Sometimes, we write the response using a writer, and sometimes (like here) we don't. Is this on purpose?

Reply:
Writer if it is asynchronous stream, else not necessary.